### DISCUSSION TITLE
Training using IntegralOperator is very slow

### DISCUSSION THREAD

**Discussion by arvasu-kulkarni** [neuraloperator/neuraloperator#370] [Ideas]
Trying to code a model using the basic Integral Operator definition given in the original Neural Operator paper ([Kovachki, Li et al](https://arxiv.org/abs/2108.08481)). The implementation in ```IntegralTransform``` is very slow to train (~30 seconds per data point).

I believe this may be due to the use of ```torch.repeat_interleave``` in line 148 of ```integral_transform.py```. There are also many instances of indexing which could slow down backprop. My model architecture is:

```
class IntegralNO(nn.Module):
    def __init__(self, in_channels=2, out_channels=1, layers=2, hidden_channels=64):
        super().__init__()
        with torch.device(device):
            self.in_channels=in_channels
            self.out_channels=out_channels
            self.layers = layers
            self.hidden_channels=hidden_channels

            # the mlp for integral blocks need to be twice the width
            self.mlp_list = [nn.Linear(2 * self.hidden_channels, self.hidden_channels)] * self.layers

            self.integral_blocks = [IntegralTransform(mlp=self.mlp_list[i], mlp_non_linearity='linear') for i in range(self.layers)]
            self.function_blocks = [nn.Linear(self.hidden_channels, self.hidden_channels)] * self.layers
            
            #print(self.function_blocks)
            self.lifting_layer = nn.Linear(self.in_channels, self.hidden_channels)
            self.projection_layer = nn.Linear(self.hidden_channels, self.out_channels)
            self.output_layer = nn.Linear(self.out_channels, self.out_channels)
    
    def forward(self, y, neighbors, x=None):
        with torch.device(device):
            final_out = torch.Tensor([]).to(device)
            for idx in range(y.shape[0]):
                # passing one row (1024x2) through network
                row = y[idx]
                neighbor_dict = neighbors[idx]
                out = row
                out = self.lifting_layer(out)

                for i in range(self.layers):
                    f_y = self.function_blocks[i].forward(out)
                    out = self.integral_blocks[i](out, neighbor_dict, f_y=f_y)

                out = self.projection_layer(out)
                out = self.output_layer(out)

                # append to final output
                final_out = torch.cat([final_out, out], dim = 1)
            return final_out
```

**Comment by dhpitt** [comment]
Thanks for this issue @arvasu-kulkarni . Some more context about your use would be helpful. Are you using CUDA or CPU?

**Comment by arvasu-kulkarni** [comment]
Hi, thanks for the reply @dhpitt . I am using CUDA.

**Comment by dhpitt** [comment]
The unbatched GINO is quite slow, especially when operating in unbatched mode. It's a known problem that isn't totally solved yet. If the geometry is constant between elements in the batch, you could try running it in batched mode for a speedup. 

Is `y` initially of shape `(batch, 1024, 2)`? How large is the batch size per forward call and how large are the neighborhoods?

**Comment by arvasu-kulkarni** [comment]
I am using a batch size of 5, i increased it to 20 and saw very little (or no) improvement. ```y``` is indeed of shape ```(b, 1024, 2)```. The neighborhoods are ~5 elements each, with some fluctuation (probably a maximum of 10).

If you have some insight on why it is slow I would love to try and patch this. I have a suspicion it is due to the repeated indexing and slicing in quantities like ```num_reps``` is slowing it down.

**Comment by dhpitt** [comment]
Like I said, the slowness is a known issue. Please feel free to open a PR if you can think of a fix, it would be a huge help! There definitely could be some special cases when `x` is a subset of `y` that could lead to a big speedup.

Re: batch size, what I'm suggesting is that your forward call splits up each element in the batch and passes them through individually. If the neighborhoods are not consistent across the batch then you can't batch the transform; in that case, batch size should have no effect.

**Comment by dhpitt** [comment]
Converting to a discussion since no PR has been opened. Please feel free to suggest a fix!
